Userland/Libraries/LibCompress/Deflate.cpp

```diff
@@ -448,4 +449,571 @@ void DeflateDecompressor::decode_codes(CanonicalCode& literal_code, Optional<Can
+{
+    // block header + literal code count + distance code count + code length count
+    auto length = 3 + 5 + 5 + 4;
+
+    // 3 bits per code_length
+    length += 3 * code_lengths_count;
+
+    for (size_t i = 0; i < code_lengths_frequencies.size(); i++) {
+        auto frequency = code_lengths_frequencies[i];
+        length += code_lengths_bit_lengths[i] * frequency;
+
+        if (i == 16) { // copy extra bits
+            length += 2 * frequency;
+        } else if (i == 17) { // zeros extra bits
+            length += 3 * frequency;
+        } else if (i == 18) { // long zeros extra bits
+            length += 7 * frequency;
+        }
+    }
+
+    return length + huffman_block_length(literal_bit_lengths, distance_bit_lengths);
+}
+
+void DeflateCompressor::write_huffman(const CanonicalCode& literal_code, const CanonicalCode& distance_code)
+{
+    for (size_t i = 0; i < m_pending_symbol_size; i++) {
+        if (m_symbol_buffer[i].distance == 0) {
+            literal_code.write_symbol(m_output_stream, m_symbol_buffer[i].literal);
+            continue;
+        }
+
+        auto symbol = length_to_symbol[m_symbol_buffer[i].length];
+        literal_code.write_symbol(m_output_stream, symbol);
+        // Emit extra bits if needed
+        m_output_stream.write_bits(m_symbol_buffer[i].length - packed_length_symbols[symbol - 257].base_length, packed_length_symbols[symbol - 257].extra_bits);
+
+        auto base_distance = distance_to_base(m_symbol_buffer[i].distance);
+        distance_code.write_symbol(m_output_stream, base_distance);
+        // Emit extra bits if needed
+        m_output_stream.write_bits(m_symbol_buffer[i].distance - packed_distances[base_distance].base_distance, packed_distances[base_distance].extra_bits);
+    }
+}
+
+size_t DeflateCompressor::encode_huffman_lengths(const Array<u8, max_huffman_literals + max_huffman_distances>& lengths, size_t lengths_count, Array<code_length_symbol, max_huffman_literals + max_huffman_distances>& encoded_lengths)
+{
+    size_t encoded_count = 0;
+    size_t i = 0;
+    while (i < lengths_count) {
+        if (lengths[i] == 0) {
+            auto zero_count = 0;
+            for (size_t j = i; j < min(lengths_count, i + 138) && lengths[j] == 0; j++)
+                zero_count++;
+
+            if (zero_count < 3) { // below minimum repeated zero count
+                encoded_lengths[encoded_count++].symbol = 0;
+                i++;
+                continue;
+            }
+
+            if (zero_count <= 10) {
+                encoded_lengths[encoded_count].symbol = 17;
+                encoded_lengths[encoded_count++].count = zero_count;
+            } else {
+                encoded_lengths[encoded_count].symbol = 18;
```

Comment:
Maybe these "symbols" (0, 17, 18, ...) could use some names?
e.g. an enum?
I dunno what it's referring to though, so feel free to ignore this if it makes no sense

Reply:
Yep they do have names and theyre used in several places (including the decompressor) so ill turn them into an enum :+1: